`copilot-core`: Implement missing cases of type equality for arrays and structs. Refs #400.

RyanGlScott opened the pull request:
The `TestEquality Type` instance is used as part of the code which checks to see if an `extern` has been declared to have two different types. This instance lacked cases for `Array` and `Struct` types, which meant that `copilot` would incorrectly believe that a nested array type (e.g., `Array 2 (Array 2 Int16)` or `Array 2 StructType`) was not equal to itself, leading to `extern`s with that type erroneously being rejected.

This patch adds the missing `Array` and `Struct` cases to the `TestEquality` instance. This proves relatively straightfoward:

* `Array n1 a1` equals `Array n2 a2` iff `n1` equals `n2` and `a1` equals `a2`.
* Two `Struct` types are equal iff they are equal according to their `Typeable` instances.

There is a very similar `EqualType Type` instance that also lacks cases for `Array` and `Struct`. This patch does not change the `EqualType Type` instance, however, for two reasons:

1. The `EqualType` class is defined in the `Copilot.Core.Type.Equality` module, which is deprecated.
2. This instance is not used anywhere in the `copilot` ecosystem.

ivanperez-keera commented:
Change Manager:

Thank you @RyanGlScott for the very quick turnaround.

This is a change in `copilot-core`, not `copilot-c99`. The issue subject and the commit messages should mention `copilot-core` instead. Also, the CHANGELOG that needs to be modified is the one in `copilot-core`.

I'd say that a more precise description of the change would be something along the lines of "implement missing cases of type equality for arrays and structs".

Please, make those changes overwriting the old commits in this PR and push force to your branch. That will trigger a re-build on Travis. Let me know when it builds successfully and I'll merge it.

Thanks!

RyanGlScott commented:
My apologies, I must have had `copilot-c99` on the brain! I've corrected these references to mention `copilot-core` instead, put the changelog entry in the correct place, and changed the description to use your suggested wording.

RyanGlScott commented:
The Travis build has passed.

ivanperez-keera commented:
Change Manager: Verified that:
- Solution is implemented:
  - [X] The code proposed compiles and passes all tests. Details:
    Build log: https://app.travis-ci.com/github/Copilot-Language/copilot/builds/259392071

  - [X] The solution proposed produces the expected result. Details:
    The following Dockerfile checks that an example with an extern with nested arrays, and one with an extern with an array of structs, can both be compiled and the resulting C code can also be compiled, in which case it prints the message `Success`.
    ```Dockerfile
      --- Dockerfile-verify-400
      FROM ubuntu:trusty
      
      RUN apt-get update
      
      RUN apt-get install --yes software-properties-common
      RUN add-apt-repository ppa:hvr/ghc
      RUN apt-get update
      
      RUN apt-get install --yes ghc-8.6.5 cabal-install-2.4
      RUN apt-get install --yes libz-dev
      
      ENV PATH=/opt/ghc/8.6.5/bin:/opt/cabal/2.4/bin:$PWD/.cabal-sandbox/bin:$PATH
      
      RUN cabal update
      RUN cabal v1-sandbox init
      RUN cabal v1-install alex happy
      RUN apt-get install --yes git
      
      ADD Arrays.hs /tmp/Arrays.hs
      ADD Structs.hs /tmp/Structs.hs
      
      SHELL ["/bin/bash", "-c"]
      CMD git clone $REPO && cd $NAME && git checkout $COMMIT && cd .. \
        && cabal v1-install $NAME/copilot**/ \
        && cabal v1-exec -- runhaskell /tmp/Arrays.hs \
        && gcc -Wall -c externNestedArrays.c \
        && cabal v1-exec -- runhaskell /tmp/Structs.hs \
        && gcc -Wall -c externArrayOfStructs.c \
        && echo "Success"
      
      --- Arrays.hs
      {-# LANGUAGE DataKinds #-}
      module Main (main) where
      
      import qualified Prelude as P
      
      import Language.Copilot
      import Copilot.Compile.C99
      
      e :: Stream (Array 2 (Array 3 Int16))
      e = extern "e" Nothing
      
      spec :: Spec
      spec = trigger "t" ((e .!! 0) .!! 1 == 42) [arg e]
      
      main :: IO ()
      main = do
        spec' <- reify spec
        compile "externNestedArrays" spec'
      
      --- Structs.hs
      {-# LANGUAGE DataKinds #-}
      module Main (main) where
      
      import qualified Prelude as P
      
      import Language.Copilot
      import Copilot.Compile.C99
      
      data S = S { field :: Field "field" Int16 }
      
      instance Struct S where
        typename _ = "s"
        toValues s = [Value Int16 (field s)]
      
      instance Typed S where
        typeOf = Struct (S (Field 0))
      
      e :: Stream (Array 2 S)
      e = extern "e" Nothing
      
      spec :: Spec
      spec = trigger "t" ((e .!! 0)#field == 42) [arg e]
      
      main :: IO ()
      main = do
        spec' <- reify spec
        compile "externArrayOfStructs" spec'
    ```
    Command:
    ```
    $ docker run -e "REPO=https://github.com/GaloisInc/copilot-1" -e "NAME=copilot-1" -e "COMMIT=8a6d32f31d5976eefd0a312f2f07d522c61e219e" -it copilot-verify-400
    ```
- [X] Implementation is documented. Details:
  No documentation needed. This change is internal (a fix to an underimplemented function).
- [X] Change history is clear.
- [X] Commit messages are clear.
- [X] Changelogs are updated.
- [X] Examples are updated. Details:
  No updates needed.
- [X] Required version bumps are evaluated. Details:
  Bump not needed. This change is internal (a fix to an underimplemented function).
